fix: change `search` to use graphql new endpoint

ignaciovi opened the pull request:
As discussed in this issue: https://github.com/tomquirk/linkedin-api/issues/313#issue-1726611738, `search` endpoint has changed in Linkedin, breaking `search_people` and `search_companies` endpoints that use it.

I basically took the solutions of [17314642](https://github.com/tomquirk/linkedin-api/issues/313#issuecomment-1574333025) and [Timur-Gizatullin](https://github.com/tomquirk/linkedin-api/issues/313#issuecomment-1578338809) and modified a few things.

There are some fields that were previously extracted in `serach_people` that I couldn't find in the new elements retrieved. There might be cleaner ways to fix this issue so feel free to suggest or add any changes!

Stelminator reviewed (commented):
just a couple of cases I found trying to use this.

gursheyss commented:
can search_jobs be fixed as well

ignaciovi commented on linkedin_api/linkedin.py:
```diff
@@ -224,18 +225,50 @@ def search(self, params, limit=-1, offset=0):
-                f"/search/blended?{urlencode(default_params, safe='(),')}",
-                headers={"accept": "application/vnd.linkedin.normalized+json+2.1"},
+                f"/graphql?variables=(start:{default_params['start']},origin:{default_params['origin']},"
+                f"query:(keywords:{default_params['keywords']},flagshipSearchIntent:SEARCH_SRP,"
```
Would you check first if `keywords` exists as a key and if it does then append `keywords:{default_params['keywords']}` inside the `query()`? Thanks for the review!

Stelminator replied:
@ignaciovi sorry for the late response, that's exactly what I'd suggest.

something like:
```python
f"/graphql?variables=(start:{default_params['start']},origin:{default_params['origin']},"
"query:("
f"keywords:{default_params['keywords']}," if 'keywords' in default_params else ""
"flagshipSearchIntent:SEARCH_SRP,"
```

which I think still reads pretty clean.

tomquirk replied:
@ignaciovi is this comment resolved?

ahnafaf commented:
I've implemented this solution on to my library, this fixes the search_people() but does not fix search_jobs()

engahmed1190 commented:
@ahnafaf can you please share an example of input parameter that will be used in `search` function for `search_people`

ahnafaf commented:
> please

api = Linkedin(data["linkedin_email"], data["linkedin_password"])

search = api.search_people(keywords = "friend_name")


> @ahnafaf can you please share an example of input parameter that will be used in `search` function for `search_people`

api = Linkedin(data["linkedin_email"], data["linkedin_password"])
search = api.search_people(keywords = "friend_name")

ignaciovi commented:
> I've implemented this solution on to my library, this fixes the search_people() but does not fix search_jobs()

Correct, `search_people()` and `search_companies()` rely on the `search()` function, which is what I modified. Unfortunately, `search_jobs()` doesn't rely on it (I'm not sure if `search()` can be modified for this). It uses the following:

```
res = self._fetch(
                f"/search/hits?{urlencode(default_params, safe='(),')}",
                headers={"accept": "application/vnd.linkedin.normalized+json+2.1"},
            )
```

You could try looking at the endpoint called on Network tab when searching for jobs in Linkedin. That's what I did to figure out how to fix the `search()`

Njoselson commented:
Hey all! I am trying to look into the `search_jobs` endpoint today, but I don't see why we need to wait for that to merge this PR? What still needs to be done and who has write access to approve?

Njoselson commented:
is it @Stelminator ?

Njoselson commented:
or @tomquirk ?

Stelminator commented:
@Njoselson 
> is it @Stelminator ?

Nope. I just left a review when I encountered some trouble with the diff. Thanks for the fix!

Njoselson commented:
> Nope. I just left a review when I encountered some trouble with the diff. Thanks for the fix!

Ok thanks! I guess there must be some other people with write access than @tomquirk ... I don't have push access so I can't see collaborators but someone with access should be able to run something like:
```
gh api repos/tomquirk/linkedin-api/collaborators
```

Njoselson commented on linkedin_api/linkedin.py:
```diff
@@ -223,19 +224,59 @@ def search(self, params, limit=-1, offset=0):
             res = self._fetch(
-                f"/search/blended?{urlencode(default_params, safe='(),')}",
-                headers={"accept": "application/vnd.linkedin.normalized+json+2.1"},
+                f"/graphql?variables=(start:{default_params['start']},origin:{default_params['origin']},"
```
Looking at this PR again, I wonder if it would be more readable to do this with `urlencode` on the dict of `default_params` like is done in the [search_jobs method](https://github.com/tomquirk/linkedin-api/blob/5273fdfd4845b3eab588c08063a96f4910b560f5/linkedin_api/linkedin.py#L503-L518). This code isn't working right now obviously since they changed the API, but I think it looks cleaner than this long URI string to be passed to the `_fetch` method.

Njoselson replied:
I have an example of how I did this same kind of thing in #336

tomquirk replied:
Thanks for the suggestion @Njoselson. @ignaciovi perhaps you can consider for a follow-up PR?

tomquirk reviewed (approved):
@ignaciovi legendary PR here, thank you! ~~Merging~~

Could you confirm if @Stelminator's comments are resolved before we merge? Thanks!

tomquirk commented on linkedin_api/linkedin.py:
```diff
@@ -249,7 +282,7 @@ def search(self, params, limit=-1, offset=0):
             self.logger.debug(f"results grew to {len(results)}")
 
-        return results
+            return results
```
@ignaciovi is this comment resolved?

ignaciovi commented:
@tomquirk thanks for taking a look! I can confirm that both comments have been resolved.

The only request left comes from https://github.com/tomquirk/linkedin-api/pull/336#issuecomment-1699311849 regarding a better way of formatting the url fetched. I can't find a way of doing it in a cleaner way since the `default_params` are stored in different parts of the query (`variables`, `query` and `queryParameters`). We can get this PR merged in order to fix the issue and anyone else or myself can take a look at improving the formatting after this, if everyone is happy with that

Njoselson commented:
I definitely think that we can open an issue for this improvement as a nice to have... I can work on that next week. It could be someone's good first issue!

Otherwise this PR looks great! Good initiative!

Stelminator commented:
@tomquirk 
> Could you confirm if @Stelminator's comments are resolved before we merge? Thanks!

I haven't tested it explicitly, but the change looks correct to address the issue I ran into with the earlier version. Still better to have this merged in any case.

Luunynliny commented:
I was about to report the weird behavior of the `search_people` function but thanks to @ignaciovi and @Njoselson.
Any update on when will the PR merge and the updates available ?

tomquirk commented:
Lets do it!

Luunynliny commented:
> Lets do it!

What a god 😅.
Thanks @tomquirk for your quickness

Stelminator commented:
> @tomquirk
> 
> > Could you confirm if @Stelminator's comments are resolved before we merge? Thanks!
> 
> I haven't tested it explicitly, but the change looks correct to address the issue I ran into with the earlier version. Still better to have this merged in any case.

closing the loop on this now that I've used it again: yes, works great, thanks!
